### PR TITLE
Reworked modf

### PR DESCRIPTION
**Pull request by SeanTheGleaming**
After noticing that math.modf only supported `f32` and `f64`, I went to check the source code to genericize the function., and ended up with a function that does pretty much exactly the same thing as the current modf implementation, but generic, supporting all float types. I went to benchmark this function, comparing it to a naive version of modf which just used `@trunc` and subtracted from the result, like so:
```
pub fn modf(x: anytype) Modf(@TypeOf(x)) {
    @setFloatMode(.strict);
    var result = Modf(@TypeOf(x)){
        .ipart = @trunc(x),
        .fpart = undefined,
    };
    result.fpart = x - result.ipart;
    return result;
}
```
My full implementation had near-identical performance to the current std implementation, but I was surprised to notice that the above simple version of `modf` was getting the same, if not better performance, while also passing all tests. Upon taking these functions to compiler explorer, I discovered that the simple version of the function compiled down into 3 instructions, unlike my more advanced implementation and the current std implementation, which both entail bit manipulation and branching, and thus have much larger bodies with worse performance

Furthermore, because this simplified function has no branching and its only type-specific function is `@trunc`, this means it will work with `@Vector`s and `comptime_float` out of the box

So I've just added my implementation, revamped some of the tests, and altered `math.zig` where appropriate.
The only thing I'm unsure of is what to do with the types `modf32_result` and `modf64_result`. Since these changed make `modf` fully generic, I have introduced a function `Modf` to generate the return type, and added that definition to `math.zig`. Now I'm unsure how much `modf32_result` or `modf64_result` get used, so I am unsure whether they should be deprecated and replaced with `Modf`, or if they should be assigned to `Modf(f32)` / `Modf(f64)` within `math.zig`. In this PR, I have gone the route of deprecating them, but I am open to the alternative if anyone thinks it's a better idea.

So, to summarize the changes:
- `modf` implementation reworked
- More tests added to `modf`
- `modf` now works with all floating point types
- `modf` now works with `@Vector`s
- At comptime, `modf` now works with `comptime_float`
- `Modf` type function added to `std.math`, for the result of `modf` (eg. `modf(@as(f32, ...))` -> `Modf(f32){ ... }`)
- `modf32_result`  and `modf64_result` deprecated and replaced with `Modf(f32)` and  `Modf(f64)` respectively

### CODE REVIEW

**Comment by SeanTheGleaming** [comment]
It looks like `@trunc` for `@Vector`s isn't implemented in some targets, so tests for vectors were removed

**Comment by nektro** [comment]
you can add `return error.SkipZigTest; // TODO` behind an if statement to skip those targets and allow the others :)

File: lib/std/math/modf.zig [hunk 1]
```diff
@@ -1,207 +1,143 @@
...
-    const T = @TypeOf(x);
-    return switch (T) {
-        f32 => modf32(x),
-        f64 => modf64(x),
-        else => @compileError("modf not implemented for " ++ @typeName(T)),
+pub fn modf(x: anytype) Modf(@TypeOf(x)) {
+    @setFloatMode(.strict);
+    var result = Modf(@TypeOf(x)){
+        .ipart = @trunc(x),
+        .fpart = undefined,
```
Comment (Vexu) [inline lib/std/math/modf.zig:25]: You can avoid mutable variables here:
```suggestion
    const ipart = @trunc(x);
    return .{
        .ipart = ipart,
        .fpart = x - ipart,
    };
```

**Review by Vexu — APPROVED** [review]

File: lib/std/math/modf.zig [hunk 2]
```diff
@@ -1,207 +1,142 @@
...
 ///  - modf(+-inf) = +-inf, nan
 ///  - modf(nan)   = nan, nan
-pub fn modf(x: anytype) modf_result(@TypeOf(x)) {
-    const T = @TypeOf(x);
-    return switch (T) {
-        f32 => modf32(x),
-        f64 => modf64(x),
-        else => @compileError("modf not implemented for " ++ @typeName(T)),
+pub fn modf(x: anytype) Modf(@TypeOf(x)) {
+    @setFloatMode(.strict);
```
Comment (andrewrk) [inline lib/std/math/modf.zig:22]: Please document the reason for choosing strict float mode, otherwise it's going to get reverted.

**Review by andrewrk — CHANGES_REQUESTED** [review]
Looks good. This basically makes the function pass through to compiler-rt trunc implementation which is at lib/compiler_rt/truncf.zig, while keeping the compiler aware of the semantics of the operation. That's exactly what it should be doing.

File: lib/std/math/modf.zig (hunk 2 above)
Comment (SeanTheGleaming) [reply lib/std/math/modf.zig]: I made that choice to ensure proper behavior with zero values, but after testing without, it seems safe to remove
